MD: the about page

shanzi opened the pull request:
Here is the about page, half finished.

The page is responsive to the width of page. When it is thin:

![aboutview-thin](https://cloud.githubusercontent.com/assets/557294/7266578/93904322-e8de-11e4-8a8b-ea7846c8f411.png)

when it is fat:

![aboutview-fat](https://cloud.githubusercontent.com/assets/557294/7266582/9a5a3a96-e8de-11e4-8208-9af4995f2967.png)

I wrote a directive to displaying objects like the old `rawdata` directive, but is more elegant and clean. I omitted the nested displaying function, but using `JSON.stringify` for a better view of the value. The field can be collapsed or expanded displaying JSON in different indention.

Field collapsed:

![field-collapsed](https://cloud.githubusercontent.com/assets/557294/7266602/eaef4a14-e8de-11e4-8c30-f4fd7e5cfc7a.png)

Field expanded:

![field-expanded](https://cloud.githubusercontent.com/assets/557294/7266610/f615439e-e8de-11e4-93b6-f2ecb3ee2d43.png)

Live demo at: http://buildbot.apps.io-meter.com/#/about

**NOTE**: as displaying the api specs depend on `buildbotService`, it has not been added yet.

### TODO
- [x] tests for `inpectdata` directive
- [x] use arrows to expand and collapse fields
- [x] displaying API spec

shanzi commented:
Added tests for `inspectdata` and some styles improvements.

Followed @tothandras 's suggestions about `bindToController` for the directive.

shanzi commented:
@tothandras as suggested that all of our directives should have a prefix as namespace like `bb-`, I think it is a good idea. What do you think about this? cc @tardyp

tardyp commented:
Well, I think this is a very important for bower libraries. but for the main application, this is not really necessary.

tardyp commented on www/md_base/src/app/common/directives/inspectdata/inspectdata.directive.coffee:
```diff
@@ -0,0 +1,53 @@
+            key: k
+            value: @processValue(v)
+
+    constructor: ($scope) ->
```
we usually put the constructor first.

tardyp commented on www/md_base/src/styles/inspectdata.less:
```diff
@@ -0,0 +1,46 @@
+.inspect-data {
```
please put the styles for the directive in the same directory as the directive

tardyp commented on www/md_base/src/styles/about.less:
```diff
@@ -0,0 +1,35 @@
+.about{
+  padding: 20px;
```
That is a lot of custom styles for one page.

I'd prefer we stick with a common style

shanzi replied:
The styles for section will be moved to be a general style as I think it could be used at settings page too. 

Will make this change soon :)

tardyp commented on www/md_base/src/app/common/directives/inspectdata/inspectdata.tpl.jade:
```diff
@@ -0,0 +1,13 @@
+      div.object-value(
+        ng-if="item.value.type == 'object'",
+        ng-init="collapsed = 1",
+        ng-click="collapsed = !collapsed",
```
the problem with putting ng-click here is that you cannot easily copy paste the value, and you actually dont know that this is even clickable. So I would prefer you stick with something like this for expansion:

![image](https://cloud.githubusercontent.com/assets/109859/7278462/7d81803c-e8ca-11e4-89a1-ab3791c73562.png)

shanzi replied:
Added a sticker will make the html a little complex. I prefer not to add one. 

The mouse cursor over the field will become pointer to indicate it is clickable and I think ng-click is smart enough to distinguish drag and select with click. I met no problem trying select the text there.

shanzi replied:
Although saying that, I will try to implement what you said and see if it is workable with small modification.

tardyp replied:
I think html complexity is no argument for directives. I indeed really would like you try with the sticker

shanzi commented:
Please have a look at current animation effect of expanding and collapsing the field at http://buildbot.apps.io-meter.com/#/about 

Is it good enough? May be no animation is better? Add animation for different fields that have different height is a little complex.

sa2ajj commented on www/md_base/src/app/about/about.tpl.jade:
```diff
@@ -1 +1,37 @@
+  div.meta(flex-gt-md="33")
+    h2 Runtime informations
+    p
+      | Informations, configrations and statuses about this running instance of buildbot.
```
configurations

jaredgrubb replied:
/Information/  ... I dont think you can pluralize that word.

sa2ajj commented on www/md_base/src/app/about/about.tpl.jade:
```diff
@@ -1 +1,37 @@
+    inspect-data(data="about.projectInfo")
+    h3 Version Info
+    inspect-data(data="about.versions")
+    h3 Configrations
```
Configurations

sa2ajj commented:
It'd be great for a native English speaker to go through the text.

tardyp commented on www/md_base/src/app/common/directives/inspectdata/inspectdata.spec.coffee:
```diff
@@ -0,0 +1,103 @@
+
+    beforeEach inject injected
+
+    it 'should inspect data correctly', ->
```
nornally you should test only one thing for each test.
you can factorize in beforeeach the initialization of the directive, and then test each feature in a different 'it'

shanzi commented:
Just removed the animation and added arrows to expand and collapse the field

![arrows](https://cloud.githubusercontent.com/assets/557294/7321346/30c8cbc2-ead4-11e4-8ca2-e9e8e781c7f4.png)

shanzi commented:
Latest code has deployed to my live demo site: http://buildbot.apps.io-meter.com

The last part of the about page is displaying API spec, this will require the presence of `buildbotService`. This service is seems to be rewrite. Should I add the old one in push forward with it now? Or we wait for the new service code is ready?

tardyp commented:
[nit] Is there a way we could expand *after* the rotation?

Looks good anyway.

Yes, you'll need to take the buildbotService as-is waiting for andras to have something workable. I dont believe it would be hard to switch api. The new api should be similar principle, just done right.

shanzi commented:
Displaying api references:

![api-references](https://cloud.githubusercontent.com/assets/557294/7341449/e717c250-ecd4-11e4-9803-4d6e023ee560.png)

tardyp commented on www/md_base/src/app/common/services/buildbot/buildbot.service.coffee:
```diff
@@ -0,0 +1,203 @@
+class BuildbotService extends Factory
```
I would prefer that you symlink those files. As you have no intention to modify them for md_base, I would prefer to avoid duplication.

shanzi replied:
It seems that gulp cannot follow symlinks for folders when compiling, unless we make symlinks for the files one by one. It makes the things a little complex.

tardyp commented on www/md_base/src/app/about/about.tpl.jade:
```diff
@@ -1 +1,53 @@
+    div.api(ng-if="about.specs.length")
+      dl(ng-repeat="spec in about.specs|orderBy:'path'", ng-init="collapsed = 1")
+        dt
+          a(ng-click="collapsed = !collapsed", href="") /{{ spec.path }}
```
please always add .no-select when you add ng-click on a text thing. else this mess up with double click

shanzi replied:
`.no-select`? Is this an existing css class?

tardyp replied:
its in styles.less
```
.no-select {
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}
```

tardyp commented on www/md_base/src/styles/styles.less:
```diff
@@ -4,3 +4,7 @@
 }
+
+a {
+  text-decoration: none;
```
does not look right to remove text-decoration for all As

shanzi replied:
hmmm... many websites like Google has removed underline for all links these days. So I tried removing them and found the appearance is good. Maybe we can display underline only when `hover`?

tardyp replied:
We need to differenciate clickable from unclickable.
![image](https://cloud.githubusercontent.com/assets/109859/7348692/e4c1579a-ecf3-11e4-8a40-0efe5892e081.png)

here, "API Reference" is not clickable, "/application.spec" is clickable, but are in the same color, its hard to make a difference.

shanzi commented:
@tardyp
Done. Now links have underlines.

In my opinion, it looks less good than that without underline, the clickable indicating may not be that important as the mouse cursor will change on clickable elements after all. Maybe we should override the theme's default link color of angular-material to blue?

shanzi commented:
@tardyp 
Any more comments? If not, I think this PR is ready to be merged :)

tardyp commented:
ok. I was waiting for you to remove the [WIP] tag.

shanzi commented:
@tardyp :)

Have you received my emails about Hangouts? Maybe we should make sure we can access to the other through that.

tardyp commented:
I have tried to run the thing locally, and I see the following error in js console as well as in unit tests:
```
Uncaught Error: [$injector:modulerr] Failed to instantiate module app due to:
Error: [$injector:unpr] Unknown provider: bbSettingsServiceProvider
http://errors.angularjs.org/1.3.15/$injector/unpr?p0=bbSettingsServiceProvider
    at http://localhost:5000/scripts.js?_1430308546951:1:27804
    at http://localhost:5000/scripts.js?_1430308546951:2:13850
```

Are you sure you rebuilt?

shanzi commented:
@tardpy this maybe a problem of the symlink as the settings service in md is symlinked to the old files. It seems ok here.

tardyp commented:
ok, the problem was with the last minute symlinks. I fixed them, and verified it builds
I also changed the make prebuilt not to build md_base.
